src/common/FrameManager.ts

```diff
@@ -87,53 +86,70 @@ export class FrameManager extends EventEmitter {
-      this._onFrameDetached(event.frameId)
-    );
-    this._client.on('Page.frameStoppedLoading', (event) =>
-      this._onFrameStoppedLoading(event.frameId)
-    );
-    this._client.on('Runtime.executionContextCreated', (event) =>
-      this._onExecutionContextCreated(event.context)
-    );
-    this._client.on('Runtime.executionContextDestroyed', (event) =>
-      this._onExecutionContextDestroyed(event.executionContextId)
-    );
-    this._client.on('Runtime.executionContextsCleared', () =>
-      this._onExecutionContextsCleared()
-    );
-    this._client.on('Page.lifecycleEvent', (event) =>
-      this._onLifecycleEvent(event)
-    );
-    this._client.on('Target.attachedToTarget', async (event) =>
-      this._onFrameMoved(event)
+    this.setupEventListeners(this._client);
+  }
+
+  private setupEventListeners(session: CDPSession) {
+    session.on('Page.frameAttached', (event) => {
+      this._onFrameAttached(event.frameId, event.parentFrameId, session);
+    });
+    session.on('Page.frameNavigated', (event) => {
+      this._onFrameNavigated(event.frame);
+    });
+    session.on('Page.navigatedWithinDocument', (event) => {
+      this._onFrameNavigatedWithinDocument(event.frameId, event.url);
+    });
+    session.on(
+      'Page.frameDetached',
+      (event: Protocol.Page.FrameDetachedEvent) => {
+        this._onFrameDetached(event.frameId, event.reason as Protocol.Page.FrameDetachedEventReason);
+      }
     );
+    session.on('Page.frameStoppedLoading', (event) => {
+      this._onFrameStoppedLoading(event.frameId);
+    });
+    session.on('Runtime.executionContextCreated', (event) => {
+      this._onExecutionContextCreated(event.context);
+    });
+    session.on('Runtime.executionContextDestroyed', (event) => {
+      this._onExecutionContextDestroyed(event.executionContextId);
+    });
+    session.on('Runtime.executionContextsCleared', () => {
+      this._onExecutionContextsCleared(session);
+    });
+    session.on('Page.lifecycleEvent', (event) => {
+      this._onLifecycleEvent(event);
+    });
+    session.on('Target.attachedToTarget', async (event) => {
+      this._onAttachedToTarget(event);
+    });
+    session.on('Target.detachedFromTarget', async (event) => {
+      this._onDetachedFromTarget(event);
+    });
   }
 
-  async initialize(): Promise<void> {
+  async initialize(client: CDPSession = this._client): Promise<void> {
     const result = await Promise.all([
-      this._client.send('Page.enable'),
-      this._client.send('Page.getFrameTree'),
+      client.send('Page.enable'),
+      client.send('Page.getFrameTree'),
     ]);
 
     const { frameTree } = result[1];
     this._handleFrameTree(frameTree);
     await Promise.all([
-      this._client.send('Page.setLifecycleEventsEnabled', { enabled: true }),
-      this._client
+      // OOP iframes will be waiting for the debugger initially.
+      client.send('Runtime.runIfWaitingForDebugger'),
```

Comment:
Hm, I am sure I have left a comment but it's not here now:

why do we need to wait for debugger? It seems to be breaking FF tests and the code branch does not look specific to OOPIFs?

Reply:
I had to set `waitForDebuggerOnStart` to `true` in `Target.setAutoAttach` to make sure I can do the setup before anything happens that I would otherwise miss.
Not sure about FF support for this - maybe @whimboo has an idea?

Reply:
But that'd run for all targets, right? Do we have waitForDebuggerOnStart for the main target in Puppeteer?

Reply:
In our partial CDP implementation we do not support any command related to debugging, just the important pieces related to (test) automation. As such `Runtime.runIfWaitingForDebugger` is not available.

Reply:
Thanks @OrKoN for making me check again. It turned out that we don't need `waitForDebuggerOnStart` after all :)
I removed it and the tests are now passing in FF too.